File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result_set = set()
        n = len(nums)

        for first_p in range(n):
            second_p = first_p + 1
            third_p = n-1

            while second_p < third_p:
                result =  nums[first_p] + nums[second_p] + nums[third_p]
                if result == 0:
                    result_set.add((nums[first_p], nums[second_p], nums[third_p]))
                    second_p += 1
                    third_p -= 1
                elif result < 0:
                    second_p += 1
                else:
                    third_p -= 1

        return [list(triplet) for triplet in result_set]                      



        

File: 0015-3sum/0015-3sum.dart
class Solution {
  List<List<int>> threeSum(List<int> nums) {
    List<List<int>> result = [];
    nums.sort();
    Set<String> seen = {};

    for(int i = 0 ; i < nums.length - 2 ; i++) {
        int left = i + 1;
        int right = nums.length - 1;

        while(left < right) {
            int sum = nums[i] + nums[left] + nums[right];
            String key = '${nums[i]},${nums[left]},${nums[right]}';
            if(sum == 0) {
                if(!seen.contains(key)) {
                  seen.add(key);
                  result.add([nums[i], nums[left], nums[right]]);
                }
                left++;
                right--;
            }
            else if(sum > 0) {
                right--;
            }
            else {
                left++;
            }
        }
    }
   return result;
  }
}